infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/InconsistentRuntimesDetector.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes;
+
+import static java.util.Collections.emptyMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.WorkspaceRuntimes;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalRuntime;
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.util.RuntimeEventsPublisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodically checks runtimes consistency and forcibly stop ones which has inconsistent state.
+ *
+ * @author Sergii Leshchenko
+ */
+@Singleton
+public class InconsistentRuntimesDetector {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InconsistentRuntimesDetector.class);
+
+  private final RuntimeEventsPublisher eventPublisher;
+  private final WorkspaceRuntimes workspaceRuntimes;
+
+  @Inject
+  public InconsistentRuntimesDetector(
+      RuntimeEventsPublisher eventPublisher, WorkspaceRuntimes workspaceRuntimes) {
+    this.eventPublisher = eventPublisher;
+    this.workspaceRuntimes = workspaceRuntimes;
+  }
+
+  @ScheduleDelay(
+      delayParameterName = "che.infra.kubernetes.runtimes_concistency_check_period_min",
+      initialDelayParameterName = "che.infra.kubernetes.runtimes_concistency_check_period_min",
+      unit = TimeUnit.MINUTES)
+  public synchronized void check() {
+    LOG.debug("Runtimes consistency checks is running");
+    for (String runningWorkspaceId : workspaceRuntimes.getRunning()) {
```

Comment:
for me looks like a classic example where the parallel stream should be used, no? or we saving threads :-) ?

Reply:
I have feeling, that the longer process is, the less actual data here - some of WS are about to stop etc and when their turn to check it- they may already became stopped etc. This will produce some amount of errors. Maybe im wrong.

Reply:
@mshaposhnik Not sure about a parallel stream. I think it would be better to run threads in the controlled limited executor. Sounds meaningful but I would leave it as is and rework later if we will face issues with unclear errors.

> when their turn to check it- they may already became stopped etc.

Makes sense. I'm adding runtime status checking directly before abnormal stopping of runtime.